hydromt_wflow/wflow.py

```diff
@@ -2715,6 +2716,267 @@ def setup_precip_forcing(
+            # TODO:
+            # da_precip = self.data_catalog.get_geodataset(...) using geom_buffered
+            # df_precip = pd.DataFrame(...)
+            # precip_stations_fn = gpd.GeoDataFrame(...)
+            raise NotImplementedError("GeoDataset source not yet implented.")
+        else:
+            raise ValueError(f"Data source {precip_fn} not recognized.")
+
+        df_precip = df_precip.astype("float32")
+
+        # Check number of stations, at least 3 needed to interpolate spatially
+        if (len(df_precip.columns) < 3) & (interp_type != "nearest"):
+            raise ValueError(
+                f"""Only {len(df_precip.columns)} found with timeseries, which is \
+not sufficient for interpolation using {interp_type} (3 or more required)."""
+            )
+
+        # Transform station coordinates to model crs
+        gdf_stations = gdf_stations.to_crs(self.crs)
+
+        # Use model resolution for the interpolated grid
+        if not np.allclose((np.abs(self.res)), 1e-6):
+            self.logger.info(
+                f"""Mismatch in horizontal and vertical resolution ({self.res}), \
+using finest resolution during interpolation by MetPy."""
+            )
+        self.logger.info(f"""Starting interpolation of data: {df_precip.shape[1]} \
+stations and {df_precip.shape[0]} timesteps.""")
+        
+        # Calling interpolation workflow
+        precip = workflows.forcing.spatial_interpolation(
+            forcing=df_precip,
+            stations=gdf_stations,
+            interp_type=interp_type,
+            hres=min(np.abs(self.res)),
+            logger=self.logger,
+            **metpy_kwargs,
+        )
+
+        # Include model CRS and rename coordinates to match model
+        precip.raster.set_crs(self.crs)
+        precip = precip.rename(
+            {"x": self.grid.raster.x_dim, "y": self.grid.raster.y_dim}
+        )
+
+        # Expand grid dimensions when a (1 X 1 X time) result is returned
+        if precip.size == len(df_precip):
+            precip = (
+                precip.squeeze()
+                .expand_dims(
+                    {
+                        self.grid.raster.y_dim: self.grid.raster.ycoords,
+                        self.grid.raster.x_dim: self.grid.raster.xcoords,
+                    }
+                )
+                .transpose("time", self.grid.raster.y_dim, self.grid.raster.x_dim)
+            )
+
+        # Compare coverage to model extent to see if some values are missing
+        precip = precip.raster.reproject_like(self.grid[self._MAPS["elevtn"]])
+        if box(*self.bounds).contains(box(*precip.raster.bounds)):
+            # Try to stick to original interpolation method
+            fill_method = (
+                    interp_type if (interp_type in ["linear", "nearest", "cubic"])
+                    else "rio_idw"
+                )
+            precip = precip.raster.interpolate_na(
+                method=fill_method, extrapolate=True
+            )
```

Comment:
This info can be moved to interp function. This will avoid repeat if we want to implement for temperature/pet. It also keeps the alreday long wflow.py file a little more lean.